File: CODE/Codeforces Round #694 (Div. 2)/d.go
package main

import (
	"bufio"
	. "fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	const N int = 1e6
	lpf := [N + 1]int{1: 1}//定义长度为1e6 + 1的数组, lpf[1]设为1
	//lpf为最大质因数的缩写
	for i := 2; i <= N; i++ {
		if lpf[i] == 0 {
			for j := i; j <= N; j += i {
				if lpf[j] == 0 {
					lpf[j] = i
				}
			}
		}
	}
	core := func(x int) int {
		c := 1
		for x > 1 {
			p := lpf[x]
			e := 1
			for x /= p; lpf[x] == p; x /= p {
				e ^= 1
			}
			if e == 1 {
				c *= p
			}
		}
		return c
	}
	var t, n, q int
	var w int64
	for Fscan(in, &t); t > 0; t-- {
		cnt := map[int]int{}
		for Fscan(in, &n); n > 0; n-- {
			var la int
			Fscan(in, &la)
			cnt[core(la)] ++
		}
		max_0, max_1 := 0, cnt[1]
		for v, c := range cnt {
			max_0 = max(max_0, c)
			if v > 1 && c & 1 == 0 {
				max_1 += c
			}
		}
		for Fscan(in, &q); q > 0 ; q-- {
			if Fscan(in, &w); w == 0 {
				Fprintln(out, max_0)
			} else {
				Fprintln(out, max(max_0, max_1))
			}
		}
	}
}

func max(x int, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}

File: CODE/Codeforces Round #694 (Div. 2)/d.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 1e6 + 10;

typedef long long ll;

int lpf[N];

void pre() {
	for(int i = 2; i <= 1000000; ++i) {
		if(lpf[i] == 0) {
			for(int j = i; j <= 1000000; j += i) {
				if(lpf[j] == 0) {
					lpf[j] = i;
				}
			}
		}
	}
}

int core(int x) {
	int c = 1;
	while(x > 1) {
		int p = lpf[x];
		int e = 1;
		for(x /= p; lpf[x] == p; x /= p) {
			e ^= 1;
		}
		if(e == 1) {
			c *= p;
		}
	}
	return c;
}


int main() {
	pre();
	int t;
	cin >> t;
	while(t--) {
		int n;
		scanf("%d", &n);
		unordered_map <int, int> mp;
		for(int i = 1; i <= n; ++i) {
			int la;
			scanf("%d", &la);
			mp[core(la)]++;
		}
		int max_0 = 0;
		int max_1 = mp[1];
		for (auto i: mp) {
			max_0 = max(max_0, i.second);
			if (i.first > 1 && (i.second & 1) == 0) {
				max_1 += i.second;
			}
		}
		int q;
		scanf("%d", &q);
		while(q--) {
			ll w;
			scanf("%lld", &w);
			if(w == 0) printf("%d\n", max_0);
			else printf("%d\n", max(max_0, max_1));
		}
	}
	return 0;
}